base/compiler/tfuncs.jl

```diff
@@ -1015,8 +1015,17 @@ function _fieldtype_tfunc(@nospecialize(s), exact::Bool, @nospecialize(name))
     exact = exact && !has_free_typevars(s)
     u = unwrap_unionall(s)
     if isa(u, Union)
-        return tmerge(_fieldtype_tfunc(rewrap(u.a, s), exact, name),
-                      _fieldtype_tfunc(rewrap(u.b, s), exact, name))
+        ta0 = _fieldtype_tfunc(rewrap(u.a, s), exact, name)
+        tb0 = _fieldtype_tfunc(rewrap(u.b, s), exact, name)
+        ta, exacta, _, istypea = instanceof_tfunc(ta0)
+        tb, exactb, _, istypeb = instanceof_tfunc(tb0)
+        if exact && exacta && exactb
+            return Const(Union{ta, tb})
+        end
+        if istypea && istypeb
+            return Type{<:Union{ta, tb}}
+        end
+        return tmerge(ta0, tb0)
```

Comment:
```suggestion
        return Any
```

Reply:
Yes, that would certainly be sound, but might loose a bit too much precision? In particular, e.g. this test then fails:
https://github.com/JuliaLang/julia/blob/61b6994884642f8fc7e18844d37dc70b9dfdb9ab/test/compiler/inference.jl#L679

Reply:
Oh, I missed that Jeff had committed that suggestion. Should we turn that test into a `@test_broken` or is there something which is both sound and sufficiently precise?

Reply:
Maybe we want `(istypea || ta === Union{})`? I think that's the only remaining case here we would want to handle with `tmerge`?

Reply:
I think Jameson's proposal (https://github.com/JuliaLang/julia/pull/41667#discussion_r674232647) is quite an elegant solution.